src/compas_timber/elements/fasteners/fastener.py

```diff
@@ -0,0 +1,182 @@
+from compas_model.elements import Element
+from compas_model.elements import reset_computed
+
+
+class Fastener(Element):
+    """
+    A class to represent timber fasteners (screws, dowels, brackets).
+
+    Parameters
+    ----------
+    elements : list(:class:`~compas_timber.parts.Element`)
+        The elements that are connected with this fastener.
+
+    Attributes
+    ----------
+    frame : :class:`~compas.geometry.Frame`
+        The coordinate system (frame) of this fastener.
+    elements : list(:class:`~compas_timber.parts.Element`)
+        The elements that are connected with this fastener.
+
+    """
+
+    OPPOSING_SIDE_MAP = {
+        0: 2,
+        2: 0,
+        1: 3,
+        3: 1,
+        4: 5,
+        5: 4,
+    }
+
+    @property
+    def __data__(self):
+        data = super(Fastener, self).__data__
+
+        return data
+
+    def __init__(self, elements, **kwargs):
+        super(Fastener, self).__init__(elements, **kwargs)
+        self.elements = elements
+        self.features = []
+        self.attributes = {}
+        self.attributes.update(kwargs)
+        self.debug_info = []
+
+    def __repr__(self):
+        # type: () -> str
+        element_str = ["{} {}".format(element.__class__.__name__, element.key) for element in self.elements]
+        return "Fastener({})".format(", ".join(element_str))
+
+    # ==========================================================================
+    # Computed attributes
+    # ==========================================================================
+
+    @property
+    def is_fastener(self):
+        return True
+
+    @property
+    def shape(self):
+        # type: () -> Brep
+        assert self.frame
```

Comment:
I guess you need to define a way for getting the frame for the fastner from the two (or more) interacting elements.